curve25519-dalek/src/backend/serial/risc0/field.rs

```diff
@@ -0,0 +1,182 @@
+    fn add_assign(&mut self, _rhs: &'b FieldElementR0) {
+        let result = self.0.add_mod(&_rhs.0, &P);
+        self.0 = result
+    }
+}
+
+impl<'a, 'b> Add<&'b FieldElementR0> for &'a FieldElementR0 {
+    type Output = FieldElementR0;
+    fn add(self, _rhs: &'b FieldElementR0) -> FieldElementR0 {
+        let mut output = *self;
+        output += _rhs;
+        output
+    }
+}
+
+impl<'b> SubAssign<&'b FieldElementR0> for FieldElementR0 {
+    fn sub_assign(&mut self, _rhs: &'b FieldElementR0) {
+        let result = self.0.sub_mod(&_rhs.0, &P);
+        self.0 = result
+    }
+}
+
+impl<'a, 'b> Sub<&'b FieldElementR0> for &'a FieldElementR0 {
+    type Output = FieldElementR0;
+    fn sub(self, _rhs: &'b FieldElementR0) -> FieldElementR0 {
+        let mut output = *self;
+        output -= _rhs;
+        output
+    }
+}
+
+impl<'b> MulAssign<&'b FieldElementR0> for FieldElementR0 {
+    fn mul_assign(&mut self, _rhs: &'b FieldElementR0) {
+        let result = risc0::modmul_u256(&self.0, &_rhs.0, &P);
+        self.0 = result;
+    }
+}
+
+impl<'a, 'b> Mul<&'b FieldElementR0> for &'a FieldElementR0 {
+    type Output = FieldElementR0;
+    fn mul(self, _rhs: &'b FieldElementR0) -> FieldElementR0 {
+        let mut output = *self;
+        output *= _rhs;
+        output
+    }
+}
+
+impl<'a> Neg for &'a FieldElementR0 {
+    type Output = FieldElementR0;
+    fn neg(self) -> FieldElementR0 {
+        let mut output = *self;
+        output.negate();
+        output
+    }
+}
+
+impl ConditionallySelectable for FieldElementR0 {
+    fn conditional_select(
+        a: &FieldElementR0,
+        b: &FieldElementR0,
+        choice: Choice,
+    ) -> FieldElementR0 {
+        FieldElementR0(U256::conditional_select(&a.0, &b.0, choice))
+    }
+}
+
+impl FieldElementR0 {
+    /// The scalar \\( 0 \\).
+    pub const ZERO: FieldElementR0 = FieldElementR0(U256::ZERO);
+    /// The scalar \\( 1 \\).
+    pub const ONE: FieldElementR0 = FieldElementR0(U256::ONE);
+    /// The scalar \\( 2 \\).
+    pub const TWO: FieldElementR0 = FieldElementR0(U256::from_be_hex(
+        "0000000000000000000000000000000000000000000000000000000000000002",
+    ));
+    /// The scalar \\( -1 \\). Set to P - 1
+    pub const MINUS_ONE: FieldElementR0 = FieldElementR0(U256::from_be_hex(
+        "7FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEC",
+    ));
+
+    const LOW_255_BITS: U256 =
+        U256::from_be_hex("7FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF");
+
+    /// Invert the sign of this field element
+    pub fn negate(&mut self) {
+        let result = risc0::modmul_u256(&self.0, &Self::MINUS_ONE.0, &P);
+        self.0 = result;
+    }
+
+    /// Given `k > 0`, return `self^(2^k)`.
+    pub fn pow2k(&self, k: u32) -> FieldElementR0 {
+        debug_assert!(k > 0);
+        let mut z = self.square();
+        for _ in 1..k {
+            z = z.square();
+        }
+        z
```

Comment:
Updated to use denormalized multiplication. Im confused on when we need to check if the result is normalized or not?

Reply:
It can be kind of subtle, unfortunately. For example, you can have your add routine either assume that the input is normalized, or that it may be denormalized. Assuming it is normalized will probably mean a faster add routine, but require the output of multiply to be normalized so that `a + b * c` works as expected. Allowing denormalized input then avoids a normalization check in the multiplication routine.

So really you have to ask yourself, does this routine give the correct result for every representation of `a` e.g. `(a, a + p, a + 2p, ...)`. If yes, then you can pass in denormalized input. If this is true for all of your routines then you don't need normalization anywhere except when unwrapping it for e.g. serialization. If it's true for some, then you need to make sure checks exists when transitioning between where is needs to be normalized and where it doesn't.

In this case we have a routine that is repeated squaring. Squaring does not require the input to be normalized, so all the intermediate values can be denormalized. Once we return the result, it may be used in many ways, some of which may require normalized input, so it's simpler to normalize before returning.